### PR TITLE
xds/federation: resource name parsing

### PR DESCRIPTION
**Pull request by menghanl**
RELEASE NOTES: N/A

### CODE REVIEW

File: xds/internal/xdsclient/xdsresource/name.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+// Name contains the parsed component of an xDS resource name.
+type Name struct {
+	Scheme    string
+	Authority string
+	Type      string
+	ID        string
+
+	ContextParams map[string]string
```
Comment (dfawley) [inline xds/internal/xdsclient/xdsresource/name.go:35]: Can you document these fields please?

File: xds/internal/xdsclient/xdsresource/name.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// ParseName splits the name and returns a struct representation of the Name.
+//
+// If the name isn't a valid new-style xDS name, field ID is set to the input.
```
Comment (dfawley) [inline xds/internal/xdsclient/xdsresource/name.go:40]: Can you give examples of what the input might be and how it maps to the output?

Is it impossible to have an invalid input?  Or do we just return `Name.ID=input` and have the caller figure out whether there is an error?

File: xds/internal/xdsclient/xdsresource/name.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			if len(vs) > 0 {
+				// We only keep one value of each key. Behavior for multiple values
+				// is undefined.
+				ret.ContextParams[k] = vs[0]
+			}
+		}
+	}
+	// TODO: processing directive (the part comes after "#" in the URL, stored
+	// in parsed.RawFragment) is not processed. Add support for that when it's
+	// needed.
```
Comment (dfawley) [inline xds/internal/xdsclient/xdsresource/name.go:80]: Should we still store them in `Name` and use that in `String()` for future-proofing?

File: xds/internal/xdsclient/xdsresource/name_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	const (
+		a = "xdstp://auth/type/id?a=1&b=2"
+		b = "xdstp://auth/type/id?b=2&a=1"
+	)
+	aParsed := ParseName(a).String()
+	bParsed := ParseName(b).String()
+
+	if aParsed != bParsed {
+		t.Fatalf("aParsed.String() = %q, bParsed.String() =%q, want them to be the same", aParsed, bParsed)
```
Comment (dfawley) [inline xds/internal/xdsclient/xdsresource/name_test.go:104]: Nit: missing space after `=`

File: xds/internal/xdsclient/xdsresource/name.go (hunk 2 above)
Comment (menghanl) [reply xds/internal/xdsclient/xdsresource/name.go:49]: Example is in the struct godoc.

I'm leaving it to the caller to figure out if the input is valid.
For now, all invalid input returns `Scheme=""` (note that it is still a valid old-style non-federation resource name)

File: xds/internal/xdsclient/xdsresource/name.go (hunk 3 above)
Comment (menghanl) [reply xds/internal/xdsclient/xdsresource/name.go]: Done

File: xds/internal/xdsclient/xdsresource/name_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// TestNameStringCtxParamsOrder covers the case that if two names differ only in
+// context parameter __order__, the parsed name.String() has the same value.
+func TestNameStringCtxParamsOrder(t *testing.T) {
+	defer func() func() {
+		oldEnv := envconfig.XDSFederation
+		envconfig.XDSFederation = true
+		return func() { envconfig.XDSFederation = oldEnv }
+	}()()
```
Comment (dfawley) [inline xds/internal/xdsclient/xdsresource/name_test.go:94]: Is this intentionally obfuscated?

```go
oldEnv := envconfig.XDSFederation
envconfig.XDSFederation = true
defer func() { envconfig.XDSFederation = oldEnv }()
```

Is much more readable (1/3rd as many `func()`s) and 2 lines shorter.

**Review by dfawley — CHANGES_REQUESTED** [review]

**Review by dfawley — APPROVED** [review]
Wrong button. Ready to submit after changing the test code.

File: xds/internal/xdsclient/xdsresource/name_test.go (hunk 5 above)
Comment (menghanl) [reply xds/internal/xdsclient/xdsresource/name_test.go]: Not intentionally intentional, but I did want to limit the scope of the temp vars.

Updated.

**Pull request merged**